### ISSUE TITLE
Cannot use get function in task with once method

### ISSUE THREAD

**Issue by ozgurky** [deployphp/deployer#921]
| Q                 | A
| ----------------- | ---
| Issue Type        | Question
| Deployer Version  | 4.0.2
| Local Machine OS  | Ubuntu 16.04 LTS
| Remote Machine OS | Ubuntu 16.04 LTS

 I use a custom task for running artisan migrate on a lumen project and I want this task to run once. 

```
task('deploy:artisan_migrate', function () {
    run("cd {{release_path}} && php artisan migrate --force");
})->once();
```
It returns `Configuration parameter deploy_path does not exists.` error but this parameter is defined in yml file. I've also tried get('release_path') and got same error. When i remove once method it seems ok.

Is not possible to use **get** function in a task with once method?

**Comment by elfet** [comment]
What do you think `get('release_path')` should return in `once()` task?

**Comment by ozgurky** [comment]
Thanks for reply, I've multiple servers for production but they don't have any git installation. So i decided to run git on local server, compress codes with tar and upload tar file to these servers. I want to run git clone command only once for fast deployment. Here is my code:

```
task('local:update_code', function () {
    $repository = trim(get('repository'));
    $branch = get('branch');
    $git = get('bin/git');
    $gitCache = get('git_cache');
    $depth = $gitCache ? '' : '--depth 1';

    if (input()->hasOption('tag')) {
        $tag = input()->getOption('tag');
    } elseif (input()->hasOption('revision')) {
        $revision = input()->getOption('revision');
    }

    $at = '';
    if (!empty($tag)) {
        $at = "-b $tag";
    } elseif (!empty($branch)) {
        $at = "-b $branch";
    }

    if (!empty($revision)) {
        runLocally("$git clone $at --recursive -q $repository {{local_path}} 2>&1");
        runLocally("cd {{local_path}} && $git checkout $revision");
    } else {
        runLocally("$git clone $at $depth --recursive -q $repository {{local_path}} 2>&1");
    }
})->once()->desc('Updating local code');
```

branch parameter is defined in server.yml file but when i use task with once get('branch') returns empty.

Is there a way for using yml file parameters with once method?

**Comment by elfet** [comment]
```yml
prod:
    ...
    branch: prod

beta:
    branch: beta
```


What should `gert('branch')` return?

**Comment by ozgurky** [comment]
```
prod-b:
  ...
  stage: develop
  branch: develop

prod-c:
  stage: production
  branch: master
```
When I run deployer like `dep  -f=deploy.php deploy develop` it should return develop and when the command is

`dep  -f=deploy.php deploy production` it should return master or do i completely misunderstand.

**Comment by elfet** [comment]
And for this? `dep deploy develop`

```
prod-a:
  ...
  stage: develop
  branch: feature

prod-b:
  ...
  stage: develop
  branch: develop

prod-c:
  stage: production
  branch: master
```

**Comment by null9beta** [comment]
@elfet  I currently got a similar problem and would need to thought on that

Sample Codes below.
The 'deploy:remote:sync' shall run on all servers but the git stuff shall only be executed once locally.

The problem now is that tasks defined with once, do not have access to any set values.
I totally got your reasoning and arguments about the get('branch') thing from above.

But any custom stuff set could be available.

What could a do achive that? What would be your suggestion?

And on thought from my side: i already loaded that config outside of a task, but only inside the tasks the input() method and other stuff is available. So putting the whole bootstrap completely outside and using e.g Constants was not working either.

```
desc('Deploy ...');
task('deploy', [

    'deploy:bootstrap',

    # git stuff
    'deploy:git:update_git',
    'deploy:git:checkout_git',

    ..
    'deploy:remote:sync',
    ..    

]);
```

```
desc('Boostrap');
task('deploy:bootstrap', function () {

    $configFile = __DIR__ . '/../config.yml';
    try {
        $config = Yaml::parse(file_get_contents($configFile));
    } catch (\RuntimeException $e) {
        throw new \RuntimeException("Error parsing {$configFile}.");
    }

    set('local_git_cache_folder', $config['something'] . get('instance_name'));
    set('otherthing', ...);
});
```

```
desc('Updating git_cache');
task('deploy:git:update_git', function () {
    runLocally('cd ' . get('local_git_cache_folder') .' && git checkout -q master && git fetch --tags --prune -q --force && git pull --all --tags -q --force');
})->once();
```

**Comment by null9beta** [comment]
@elfet and actually another problem i face just now is related to the slack recipe in the recipes repo:

1) i am asking for the branch do deploy in one task.
2) because i don´t want the task to be executed multiple times i first tried with once(), but due to the same reason described above that did not work really.
3) then i went with using the `onlyOn([...])` on the task and added only one server, which kinda solved it so far, because i was asked only once.
4) i then set the branch using `set('branch', $selectedBranch);` in the task that is only executed onlyOn ...
5) after the deployment i am using the slack recipe to send a slack message. 

here comes the problem.
The slack recipe is using `get('branch')` which then fails because the set('branch') is executed inside the task that is run onlyOn.

How would i overcome that problem?

I know i could set the branch as an argument but that`s not really elegant because in our rather big setup there is just to much branches to remember, so it´s nicer to ask for the branch showing a list of the branches (i actually use the ask() feature of the deployer).

Talking about the ask feature btw. there is another problem because asking only works with not using parallel execution (-p). Which is also a little unfortunate.

+ please don´t get that wrong, it´s a nice project that you got here and we are just about to start using it, so i am facing a few obstacles on the road now switching our old deployment to deployer.

**Comment by lnesi** [comment]
It would be nice to have this option and be more supportive of the community if not whats the point of having open source code.

**Comment by elfet** [comment]
Sorry, closed by mistake while clearing issues. 

@null9beta i see the problem, but for now i don't see elegant solution for this.
Problem is what config options server specified. Here an example:

<table>
<thead>
<tr>
<th>Server A</th>
<th>Server B</th>
<th>Server C</th>
</tr>
</thead>
<tbody>
<tr>
<td>Task 1</td>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>Task 1</td>
<td></td>
</tr>
<tr>
<td></td>
<td></td>
<td>Task 1</td>
</tr>
<tr>
<td colspan="3" align="center">Task 2 with `once` option</td>
</tr>
<tr>
<td>Task 3</td>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>Task 3</td>
<td></td>
</tr>
<tr>
<td></td>
<td></td>
<td>Task 3</td>
</tr></tbody></table>

What config options should get *Task 2*? 

> Talking about the ask feature btw. there is another problem because asking only works with not using parallel execution (-p). Which is also a little unfortunate.

This is parallel compromise. 🤗 You can't simple get stdin from background workers (or with difficult proxy on pipes). There may be some solution for this problem:

<table>
<thead>
<tr>
<th>Server A</th>
<th>Server B</th>
<th>Server C</th>
</tr>
</thead>
<tbody>
<tr>
<td>Task 1</td>
<td>Task 1</td>
<td>Task 1</td>
</tr>
<tr>
<td colspan="3" align="center">`ask()` in task with *once*</td>
</tr>
<tr>
<td>Task 3</td>
<td>Task 3</td>
<td>Task 3</td>
</tr></tbody></table>

But still, there is needed some sort of passing configuration options back to server's config.

**Comment by lnesi** [comment]
HI 
I Resolved this by: (Example)
1 Create variables needed to hold credentials on deploy.php 
$git_user='';
$git_password='';

2. Create set with function 
set('git_user',function(){
   global $git_user;
   return $git_user;
});
set('git_password',function(){
    global $git_password;
    return $git_password;
});

3. Create a task get credentials that runs once.

task('get_credentials',function(){
    global $git_user,$git_password;
   $git_user=ask('Please provide git user:'); 
   $git_password=askHiddenResponse('Please provide git password:'); 
   
})->once();

3. Where i need the credentials i use the global variables example for setting repository URL

set('repository', function(){
    global $git_user,$git_password;
    return 'https://'.$git_user.':'.$git_password.'@bitbucket.org/......';
});

Hope this helps.

**Comment by antonmedv** [comment]
Now in Deployer v5 you can do some crazy stuff:
```php

task('test', ['set', 'get', 'tie']);

task('set', function () {
    on(host('[a:f]'), function ($host) {
        $host->set('value', '{{hostname}}');
    });
})->local();

task('get', function () {
    writeln("{{hostname}}:{{value}}");
    set('key', '{{hostname}}');
});

task('tie', function () {
    $value = '';
    on(host('[a:f]'), function () use (&$value) {
        $value .= parse('{{key}}');
    });
    writeln($value);
})->local();
```

**Comment by antonmedv** [comment]
Ask for branch when deploy in parallel:
```php
task('ask_deploy', ['what_branch', 'deploy']);

task('what_branch', function () {
    $v = ask('branch?');

    on(input()->getArgument('stage'), function ($host) use ($v) {
        $host->set('branch', $v);
    });
})->local();
```
